appshell/appshell_extensions_gtk.cpp

```diff
@@ -442,44 +442,54 @@ int32 ReadFile(ExtensionString filename, ExtensionString& encoding, std::string&
     if (encoding == "utf8") {
         encoding = "UTF-8";
     }
-    int32 error = NO_ERROR;
 
-    try {
-        std::ifstream file(filename.c_str());
-        std::stringstream ss;
-        ss << file.rdbuf();
-        contents = ss.str();
-        std::string detectedCharSet;
-        try {
-            if (encoding == "UTF-8") {
-                CharSetDetect ICUDetector;
-                ICUDetector(contents.c_str(), contents.size(), detectedCharSet);
-            }
-            else {
-                detectedCharSet = encoding;
-            }
-            if (detectedCharSet == "UTF-16LE" || detectedCharSet == "UTF-16BE") {
-                return ERR_UNSUPPORTED_UTF16_ENCODING;
-            }
-            if (detectedCharSet != "UTF-8") {
-                try {
+    int error = NO_ERROR;
+    GError *gerror = NULL;
+    gchar *file_get_contents = NULL;
+    gsize len = 0;
+    
+    if (!g_file_get_contents(filename.c_str(), &file_get_contents, &len, &gerror)) {
+        error = GErrorToErrorCode(gerror);
+        if (error == ERR_NOT_FILE) {
+            error = ERR_CANT_READ;
+        }
+    } else {
+        if (has_utf_32_BOM(file_get_contents, len)) {
+            error = ERR_UNSUPPORTED_ENCODING;
+        } else  if (has_utf8_BOM(file_get_contents, len)) {
+            std::ifstream file(filename.c_str());
+            std::stringstream ss;
+            ss << file.rdbuf();
+            contents = ss.str();
+            contents.erase(0, 3);
+            preserveBOM = true;
+        } else if (!g_locale_to_utf8(file_get_contents, -1, NULL, NULL, &gerror) || encoding != "UTF-8") {
```

Comment:
Line 460 to 465 is the only change added new in this PR, rest is just revert of my last PR https://github.com/adobe/brackets-shell/pull/609/files#diff-10aa13d798996c2d964d38b1f328ae62